### PR TITLE
feat: vega wallet is now aware of core spam rules

### LINKED ISSUE #7359: Using vega wallet should never get me banned for breaking spam / pow rules
# Task Overview

Using vega wallet should never get me banned for breaking spam / pow rules

**As a** user using the command line wallet with one of our web interfaces or a developer using it to sign transactions created algorithmically
**I want** the wallet to track and increase PoW difficulty as required and check spam statistics for things like votes, delegations and proposals 
**So that** I never get banned for sending insufficiently hard PoW with my transactions and I never get banned (without a big fat warning at least) for doing too many delegations / proposals / voting too many time per proposal. 

## Specs
- [Link](xyz) https://github.com/vegaprotocol/specs/blob/master/protocol/0062-SPAM-spam_protection.md and https://github.com/vegaprotocol/specs/blob/master/protocol/0072-SPPW-spam-protection-PoW.md 

# Acceptance Criteria
How do we know when this technical task is complete:

- It is possible to use a algo trading bot with the wallet which chooses to submit arbitrarily high number of orders and all the orders get the appropriately hard PoW attached so that it's not getting banned for breaking PoW rules.
- It is possible to use the web UI to change votes arbitrarily many times on a proposal and if the key is hitting a spam limit the wallet will either refuse to send the transaction that would lead to a ban or at least give the user a very big red warning. 
- It is possible to use change delegations in the web UI many times in an epoch and if the key is hitting a spam limit the wallet will either refuse to send the transaction that would lead to a ban or at least give the user a very big red warning. 

# Additional Details (optional)
See https://github.com/vegaprotocol/vega/pull/7275

**Before Merging**
- [ ] Create relevant for [system-test](https://github.com/vegaprotocol/system-tests/issues) tickets with feature labels
- [ ] Code refactored to meet SOLID and other code design principles
- [ ] Code is compilation error, warning, and hint free
- [ ] Carry out a basic happy path end-to-end check of the new code
- [ ] All acceptance criteria confirmed to be met, or, reasons why not discussed with the engineering leadership team
- [ ] All APIs are documented so auto-generated documentation is created
- [ ] All Unit, Integration and BVT tests are passing
- [ ] Implementation is peer reviewed (coding standards, meeting acceptance criteria, code/design quality)
- [ ] Create [front end](https://github.com/vegaprotocol/token-frontend/issues) or [console](https://github.com/vegaprotocol/console/issues) tickets with feature labels (should be done when starting the work if dependencies known i.e. API changes)

**After Merging**
- [ ] Move development ticket to `Done` if there is **NO** requirement for new system-tests
- [ ] Resolve any issues with broken system-tests
- [ ] Create [documentation](https://github.com/vegaprotocol/documentation/issues) tickets with feature labels if functionality has changed, or is a new feature

### PR DESCRIPTION
**Pull request by wwestgarth**
closes #7359 

The vega wallet now uses the `GetSpamStatistics` core API instead of `LastBlockAndHash` to generate the required proof-of-work for a transaction.

The V2 API also using the max-per-epoch counts for each transaction type to prevent sending in a transaction that will trigger a spam violation in core and cause the party to be banned.

I temporaily added the new check into the V1 API and ran the smoke-tests and spam-test suites -- the smoke-tests passed and the spam tests failed in expected ways.

I have now removed the spam check from the V1 API because it is deprecated and to avoid the trouble of the spam tests breaking. I can wire it back in again if @vegaprotocol/qa think it will be useful to partially unblock testing until their V2 integration work is done.

### CODE REVIEW

File: protos/sources/vega/api/v1/core.proto [hunk 1]
```diff
@@ -405,6 +407,14 @@ message PoWBlockState {
   optional uint64 expected_difficulty = 4;
   // The hashing function used to calculate the block hash
   string hash_function = 5;
+  // The base difficulty for this block for when transactions seen < tx_per_block
+  uint64 difficulty = 6;
```
Comment (ze97286) [inline protos/sources/vega/api/v1/core.proto:411]: why do we need these 3 params here? i.e. why is expected_difficulty insufficient?

Comment (ze97286) [reply protos/sources/vega/api/v1/core.proto:411]: so if I'm not mistaken you're not even using expected_difficulty in which case maybe it's not worth bothering with it in the first place.

Comment (wwestgarth) [reply protos/sources/vega/api/v1/core.proto:411]: I know, and I asked for it to be added, so thats all on me.

File: wallet/api/spam/pow.go [hunk 2]
```diff
@@ -78,53 +87,42 @@ func (t *txCounter) resize(n int64) {
...
 }
 
 // Generate returns a proof-of-work with difficult that respects the history of transactions sent in against a particular block.
-func (p *ProofOfWork) Generate(pubKey string, lastBlock *nodetypes.LastBlock) (*commandspb.ProofOfWork, error) {
-	p.mu.Lock()
-	counter := p.getCounterForChain(lastBlock.ChainID)
+func (s *Handler) GenerateProofOfWork(pubKey string, st *nodetypes.SpamStatistics) (*commandspb.ProofOfWork, error) {
+	s.mu.Lock()
+	counter := s.getCounterForChain(st.ChainID)
+	blockState := st.PoW.PowBlockStates[0]
```
Comment (ze97286) [inline wallet/api/spam/pow.go:101]: is this the last block or some random block?

Comment (wwestgarth) [reply wallet/api/spam/pow.go:101]: Should be the latest, we sort them in `insecure_grpc_adapter.go`.

Comment (ze97286) [reply wallet/api/spam/pow.go:101]: I wonder if we want to do something smarter, like try to find the block that minimises the difficulty but anyways, the last block would be fine most of the time (it won't if suddenly there's an extra long block)

**Review by ze97286 — COMMENTED** [review]
does this PR not break all/most/some spam/pow tests in nightly?

**Review by ze97286 — DISMISSED** [review]

**Comment by wwestgarth** [comment]
> does this PR not break all/most/some spam/pow tests in nightly?

It doesn't because its not wired into the V1 API (but it would if it was). I think the plan might be to hold of merging this in until the V2 integration is a bit futher along...

**Comment by wwestgarth** [comment]
I think this is ready now. On a hacky system test branch I've got the smoke run mostly green using the walletV2 and this core branch: https://jenkins.ops.vega.xyz/blue/organizations/jenkins/common%2Fsystem-tests/detail/system-tests/17072/pipeline/
The failing tests are vegawallet tests trying to do things that you can no longer do via the V2 api i.e taintkey, add key metadata, or it returning a different error string. So I think this change is good.

**Review by ze97286 — DISMISSED** [review]

File: protos/sources/vega/api/v1/core.proto [hunk 3]
```diff
@@ -413,6 +423,8 @@ message PoWStatistic {
   repeated PoWBlockState block_states = 1;
   // PoW banned until timestamp as RFC3339Nano
   optional string banned_until = 2;
+  // The number of block behind the current block whose hash can be used for prro-of-work calculations
```
Comment (ValentinTrinque) [inline protos/sources/vega/api/v1/core.proto:426]:
```suggestion
  // The number of block behind the current block whose hash can be used for proof-of-work calculations
```

File: wallet/api/client_send_transaction_test.go [hunk 4]
```diff
@@ -29,8 +29,9 @@ func TestSendTransaction(t *testing.T) {
 	t.Run("Interrupting the request does not send the transaction", testInterruptingTheRequestDoesNotSendTransaction)
 	t.Run("Getting internal error during the review does not send the transaction", testGettingInternalErrorDuringReviewDoesNotSendTransaction)
 	t.Run("No healthy node available does not send the transaction", testNoHealthyNodeAvailableDoesNotSendTransaction)
-	t.Run("Failing to get the last block does not send the transaction", testFailingToGetLastBlockDoesNotSendTransaction)
+	t.Run("Failing to get the last block does not send the transaction", testFailingToGetSpamStatsDoesNotSendTransaction)
```
Comment (ValentinTrinque) [inline wallet/api/client_send_transaction_test.go:32]:
```suggestion
	t.Run("Failing to get the spam statistics does not send the transaction", testFailingToGetSpamStatsDoesNotSendTransaction)
```

File: wallet/api/node/retrying_node.go [hunk 5]
```diff
@@ -58,6 +59,32 @@ func (n *RetryingNode) Statistics(ctx context.Context) (nodetypes.Statistics, er
...
+	resp, err := n.grpcAdapter.SpamStatistics(ctx, pubKey)
+	if err != nil {
+		n.log.Error("could not get the statistics",
+			zap.String("host", n.grpcAdapter.Host()),
+			zap.Error(err),
+		)
+		return nodetypes.SpamStatistics{}, err
+	}
+
+	// TODO log it all....
```
Comment (ValentinTrinque) [inline wallet/api/node/retrying_node.go:74]: TODO left

File: wallet/api/node/retrying_node.go [hunk 6]
```diff
@@ -58,6 +59,32 @@ func (n *RetryingNode) Statistics(ctx context.Context) (nodetypes.Statistics, er
...
+	n.log.Debug("response from SpamStatistics",
+		zap.String("host", n.grpcAdapter.Host()),
+		zap.String("chain-id", resp.ChainID),
+		zap.Uint64("epoch", resp.EpochSeq),
+		zap.Uint64("block-height", resp.LastBlockHeight),
+		zap.Uint64("prosposals", resp.Proposals.CountForEpoch),
+		zap.Uint64("transfers", resp.Transfers.CountForEpoch),
+		zap.Uint64("delegations", resp.Delegations.CountForEpoch),
```
Comment (ValentinTrinque) [inline wallet/api/node/retrying_node.go:82]: Can we add `-count-for-epoch` behind those fields ?

**Review by ValentinTrinque — APPROVED** [review]

**Pull request merged**